File: Javascript/Sorting/Quick_Sort.js


function QuickSortPivot(arr,low,high){
    let pivot=arr[low];
    let i=low;
    let j=high;
    
    while(i<j){
      
      while(arr[i]<=pivot && i<=high-1){
        i++;
      }
      while(arr[j]>pivot && j>=low+1){
        j--;
      }
    if(i<j){
     swap(arr,i,j)
    }
    
    }
   swap(arr,low,j)
    return j;
  }
  
  
  // swaping 
  
  function swap(arr,i,j){
    let temp=arr[i];
    arr[i]=arr[j];
    arr[j]=temp;
  }
  
  
  
  function QuickSort(arr,low,high){
    if(low<high){
      
      let pivot=QuickSortPivot(arr,low,high);
      //console.log(pivot)
      QuickSort(arr,low,pivot-1);
      QuickSort(arr,pivot+1,high);
    }
  }
  
  
  
  
  function main(){
    let arr=[1,5,6,7,2,3,4];
    let n=7;
    QuickSort(arr,0,n-1)
    console.log(arr);
    
  }
  
  main()
  
  
  

File: C++/Sorting/Quick_Sort.cpp
#include <iostream>
#include <algorithm>
#include <bits/stdc++.h>
using namespace std;


int partition(vector<int>&arr, int low ,int high){
  int pivot=arr[low];
  int i=low;
  int j=high;
  
  while(i<j){
    while(arr[i]<=pivot && i<=high-1){
      i++;
    }
    while(arr[j]>pivot && j>=low+1){
      j--;
    }
    if(i<j){
      swap(arr[i],arr[j]);
    }
  }
  swap(arr[low],arr[j]);
  return j;
}

void  qs(vector<int>&arr ,int low, int high){
  if(low<high){
    int pIndex=partition(arr,low ,high);
    qs(arr,low,pIndex-1);
    qs(arr,pIndex+1,high);
  }
}

int main() 
{
   int n=7;
   vector<int>arr{1,5,6,7,2,3,4};
   qs(arr,0,n-1);
   for(int i=0;i<n;i++){
     std::cout << arr[i] << std::endl;
   }
    return 0;
}